File: partition_list/PartitionList.py
import sys
sys.path.append("/Users/jinzhao/leetcode/")
from leetcode.common import *

class Solution:
    # @param {ListNode} head
    # @param {integer} x
    # @return {ListNode}
    def partition(self, head, x):
        smallerPreHead = ListNode(0)
        largerPreHead = ListNode(0)
        currentSmallerHead = smallerPreHead
        currentLargerHead = largerPreHead
        current = head
        if not head or not head.next:
            return head
        while current  :
            if current.val < x:
                currentSmallerHead.next = current
                currentSmallerHead =currentSmallerHead.next
            else:
                currentLargerHead.next = current
                currentLargerHead = currentLargerHead.next
            current = current.next
        currentSmallerHead.next = largerPreHead.next
        #need this line to avoid circle.
        currentLargerHead.next = None 
        return smallerPreHead.next

if __name__ == "__main__":
    sol = Solution()
    l1 = construct_node_from_list([1,4,3,2,5,2])
    head = sol.partition(l1, 3)
    print head

File: partition_list/PartitionList.cpp
#include "../common.cpp"
#include <iostream>


class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        
        ListNode* smallerPreHead = new ListNode(0);
        ListNode* largerPreHead = new ListNode(0);
        ListNode* currentSmallerHead = smallerPreHead;
        ListNode* currentLargerHead = largerPreHead;
        ListNode* current = head;
        
        if (!head || !head->next ) return head;
        while (current)
        {
            if (current->val < x) {
                currentSmallerHead->next = current;
                currentSmallerHead = currentSmallerHead->next;
            }
            else{
                currentLargerHead->next = current;
                currentLargerHead = currentLargerHead->next;
            }
            current = current->next;
        }
        currentSmallerHead->next = largerPreHead->next;
        //Avoid circle
        currentLargerHead->next = nullptr;
        return smallerPreHead->next;
        
    }
};


int main()
{
    Solution sol;
    std::vector<std::pair<std::vector<int>, int>> tests = {
        { {}, 0 }
        , { {1,1,1,1,1,1,1}, 0 }
        , { {1,1,1,1,1,1,1}, 1 }
        , { {1,1,1,1,1,1,1}, 2 }
        , { {1,4,3,2,5,2}, 3 }
    };
    
    for (auto& test: tests){
        
        auto input = constructListNode(test.first);
        std::cout<<"data";
        printListNode(input);
        std::cout<<" partitioned by : "<< test.second <<std::endl;
        
        auto result = sol.partition(input, test.second);
        std::cout<<"result: ";
        printListNode(result);
        std::cout << std::endl;
    }
    return 0;
    
}